Autotools to install systemd unit files.

osynge opened the pull request:
To simplify the spec file we should install as much using autotools
and as little as possible in the spec file.

Signed-off-by: Owen Synge <osynge@suse.com>

tchaikov commented:
@osynge might need rebase. the configure.ac part has some conflicts with the change of systemd_libexec_dir.

tchaikov commented:
@osynge the `Makefile.in` is already covered by the https://github.com/ceph/ceph/blob/master/.gitignore#L50 , but Makefile is deliberately left out of the list. guess this allows us to have some hand-made Makefiles in the project.

osynge commented:
@tchaikov 
Rebased to master and added the .gitignore.

Please note this will break the RPM (and probably deb) build as new files will now need to be packaged in the spec file. If these changes need to go in this patch for being accepted please let me know.

tchaikov commented:
@osynge yes, please add another commit to package them in rpm.

seems we have not packaged the systemd files for deb, so this change won't break the deb packaging. but debian switched to systemd as the default init system for jessie[0]. probably we should also ready ceph for this change. anyway, it's completely up to you, as jessie is not officially supported yet [1], AFAIK.

---
[0] http://article.gmane.org/gmane.linux.debian.devel.ctte/5453
[1] http://comments.gmane.org/gmane.comp.file-systems.ceph.devel/24572

osynge commented:
@tchaikov I will look at this tomorrow. 

I imagine all that’s required is removing some install lines.

Unfortunately I dont have good test systems for rhel.

I need to build a new upstream testing branch for SUSE reflecting this change.

liewegas commented:
I'm assuming

- we will still package sysvinit
- systemd will ignore the sysvinit file if you don't use it
- an upgraded system with sysvinit tagged daemons will still start that way, and
- ceph-deploy or ceph-disk can start marking new stuff with systemd and the two can coexist
- later a user can go rm sysvinit ; touch systemd and reboot to make the transition

?

osynge commented:
@liewegas 

* I dont want ceph to work on only systemd or only on sysVinit based systems.
* I want "make install" to provide a working system and make packaging just packaging.
* How about we make it an autotools / configure.in conditional which init system to provide defaulting to sysVinit for now?
  * At SUSE we will only be providing systemd for the near future.
  * I dont approve of users using the rm command but we can do this in packaging if thats you prefured way. 
* If upstream wants a super set of systemd and sysVinit with "make install" this is going to confuse new contributors and src based installers.
* systemd has backward compatibility to sysVinit.
  * This leads to potential user / admin confusion when you provide sysVinit and systemd, as comparability will lead to more than one way to enable and disable a service.
   * I accept that systemd should do the right thing when both methods are provided but it does not.
* systemd provides more than init management but also provides logging and other features.
  * logging definitely will also provide different packaging dependent on sysVinit and systemd
  * using systemd logging will for systemd close one of your logging bugs.
* Upstream ceph does not necessarily have to look the same as SUSE or Redhat ceph packages
  * The more similar the better for collaboration and less patches for downstream teams to maintain.
* I care about getting a SUSE rpm working on Master for our testing to foster better collaboration.

liewegas commented:
>  *  I dont want ceph to work on only systemd or only on sysVinit based systems.

agree

>  *  I want "make install" to provide a working system and make packaging just packaging.

i haven't worried about this until now, but, sure!

>  *  How about we make it an autotools / configure.in conditional which init system to provide defaulting to sysVinit for now?
>      +  At SUSE we will only be providing systemd for the near future.
>      +  I dont approve of users using the rm command but we can do this in packaging if thats you prefured way.

it seems like we have two option:

1) we install both sysvinit and systemd files.  on upgrade, both work in 
parallel, managing different daemons (based on the 
sysvinit/upstart/systemd files in /var/lib/ceph/*/*).  the admin 
explicitly opts into a new init system by adjusting those files.

2) we dictate the init system in the package.  on upgarde, the postinst 
(or equivalent) script replaces /var/lib/ceph/*/*/sysvinit files with 
systemd files.

I guess 2 is simpler... I was assuming 1 would be least disruptive but it 
is more complex.  Is it even feasible?  That's what I don't know...

>  *  If upstream wants a super set of systemd and sysVinit with "make install" this is going to confuse new contributors and src based installers.

yeah, ok, probably.

>  *  systemd has backward compatibility to sysVinit.
>      +  This leads to potential user / admin confusion when you provide sysVinit and systemd, as comparability will lead to more than one way to
>         enable and disable a service.
>      +  I accept that systemd should do the right thing when both methods are provided but it does not.

yeah

>  *  systemd provides more than init management but also provides logging and other features.
>      +  logging definitely will also provide different packaging dependent on sysVinit and systemd
>      +  using systemd logging will for systemd close one of your logging bugs.

I haven't looked at any of this yet...

>  *  Upstream ceph does not necessarily have to look the same as SUSE or Redhat ceph packages
>      +  The more similar the better for collaboration and less patches for downstream teams to maintain.

agree.  i'd like it is things were as similar/consistent at 
possible--it'll minimize user confusion and make the docs easier to 
maintain.

>  *  I care about getting a SUSE rpm working on Master for our testing to foster better collaboration.

me too!  would like to get this into master soon so that we have some time 
to shake things out before infernalis.

osynge commented:
> it seems like we have two option:
> 
> 1) we install both sysvinit and systemd files.  on upgrade, both work in 
> parallel, managing different daemons (based on the 
> sysvinit/upstart/systemd files in /var/lib/ceph/*/*).  the admin 
> explicitly opts into a new init system by adjusting those files.
> 
> 2) we dictate the init system in the package.  on upgarde, the postinst 
> (or equivalent) script replaces /var/lib/ceph/*/*/sysvinit files with 
> systemd files.

> I guess 2 is simpler... I was assuming 1 would be least disruptive but it 
> is more complex.  Is it even feasible?  That's what I don't know...

I have a couple of responses to this.

(A) systemd support in upstream will come in a major release. People can
expect things to change in major releases and changing from sysVinit to
systemd init is not much work for the user. (apart from what is this
thing! and why me! but after that its OK, especially in automated data
centres, the biggest worry cephs's small scale users who deploy from
source, but they should know enough to help them selves)

(B) The init system is typically packaged in most distributions so on
upgrade people will not need to remove the old one.

(C) At least on SUSE the rpm hooks should do the upgrade just fine, or
else its a bug, as the hooks will/should restart the daemon on upgrade.

(D) Point (C) should be true for all package based linux distributions
and if memory serves correctly also solaris.

So I have no concerns on the upgrade issue.

This leads me to think your options (1) and (2) are both too complex.

I propose

(3) I would have autotools set the spec file paths / macros with a
central switch, and have the sepc file generated accordingly, (even if
the details end up with spec file also needs to be built with a systemd
sysVinit solution) we will keep most of the branching for all build
parameters in one place and supposedly platform neutral.

Ideally we can also build packages that (ideally) depend on init system,
and provide ceph and conflict with each other. This ideally would then
allow update testing to prove my assertions (B) (C) and (D).

Best wishes

Owen

PS I only know rpm and deb reasonably well, but I believe pacman is
similar with respect to this discussion.

liewegas commented:
On Thu, 28 May 2015, Owen Synge wrote:
> (C) At least on SUSE the rpm hooks should do the upgrade just fine, or
> else its a bug, as the hooks will/should restart the daemon on upgrade.

Note that currently the deb and rpms don't restart daemons on upgrade 
(because it can be disruptive on a running cluster).  Traditionally (for 
ceph) we've made this an explicit administrator step.

But in any case,

> This leads me to think your options (1) and (2) are both too complex.
> 
> I propose
> 
> (3) I would have autotools set the spec file paths / macros with a
> central switch, and have the sepc file generated accordingly, (even if
> the details end up with spec file also needs to be built with a systemd
> sysVinit solution) we will keep most of the branching for all build
> parameters in one place and supposedly platform neutral.

I think this is the same as my #2?  The package determines the init 
system.  Which init system it is will be determined at build time based 
on the distro (e.g., el6 will be sysvinit, el7 will be systemd).

> Ideally we can also build packages that (ideally) depend on init system,
> and provide ceph and conflict with each other. This ideally would then
> allow update testing to prove my assertions (B) (C) and (D).

That sounds like a lot of work... :(  But, there will be folks to opt-out 
of systemd (e.g., on debian) that want packages.  I'm not sure if it's 
worth catering to them or not.

But... is #1 really infeasible?  Will the presence of the 'ceph' sysvinit 
script screw up the ceph-osd@.service etc files?  We successfully made 
upstart and sysvinit coexist (by using different naming, ceph-all vs 
ceph)...

sage

osynge commented:
On 05/28/2015 09:09 PM, Sage Weil wrote:
> On Thu, 28 May 2015, Owen Synge wrote:
>> (C) At least on SUSE the rpm hooks should do the upgrade just fine, or
>> else its a bug, as the hooks will/should restart the daemon on upgrade.
> 
> Note that currently the deb and rpms don't restart daemons on upgrade 
> (because it can be disruptive on a running cluster).  Traditionally (for 
> ceph) we've made this an explicit administrator step.
> 
> But in any case,
> 
>> This leads me to think your options (1) and (2) are both too complex.
>>
>> I propose
>>
>> (3) I would have autotools set the spec file paths / macros with a
>> central switch, and have the sepc file generated accordingly, (even if
>> the details end up with spec file also needs to be built with a systemd
>> sysVinit solution) we will keep most of the branching for all build
>> parameters in one place and supposedly platform neutral.
> 
> I think this is the same as my #2?  The package determines the init 
> system.  Which init system it is will be determined at build time based 
> on the distro (e.g., el6 will be sysvinit, el7 will be systemd).

Build infrastructure should in my opinion define init system not
packaging, but from a later comment I can guess why you see this as
equivalent.

I will come back to this point at the end of the email. (hopefully not
in to much detail)

>> Ideally we can also build packages that (ideally) depend on init system,
>> and provide ceph and conflict with each other. This ideally would then
>> allow update testing to prove my assertions (B) (C) and (D).
> 
> That sounds like a lot of work... :(  But, there will be folks to opt-out 
> of systemd (e.g., on debian) that want packages.  I'm not sure if it's 
> worth catering to them or not.

Note (1) this is how I see things ideally.
Note (2) I suspect it wont be as much work as you might expect, if this
is all done at the build tools level, as the packaging will just differ
by a couple of variable.

I suspect with the fuss over systemd we will not be seeing the last of
init systems impacting server side work.

> But... is #1 really infeasible?  Will the presence of the 'ceph' sysvinit 
> script screw up the ceph-osd@.service etc files?  We successfully made 
> upstart and sysvinit coexist (by using different naming, ceph-all vs 
> ceph)...

I had never considered this way of working to be honest. I have no doubt
ceph can do this I just dont think its the least effort way of doing it
for the users or the team.

Our experiences
---------------

With our transition (not seen outside the team) in SUSE the main issue
that hit us was we used systemd as a backed to sysVinit. we probably did
things in ways upstream systemd folks know how to work around but we
could enable on boot processes twice, and then when you disable them
once they dont disable.

This proved confusing to us and we imagined the support would explode so
we stopped this.

Things would have been different and potentially worse if we had 100%
parallel boot init systems in the package as you suggest. Our fear was
that OSD's and mons could potentially have attempted to start with the
same name twice. Since we knew we only had to support systemd by SUSE
policy fro SLE12 and going forward we made the call to use systemd
incrementally more and more through our packages.


How I think it would be best to move forward.
---------------------------------------------

I think as upstream project lead it is important to care about the
differences between packaging and build tools.

To me build tools should support all permutations to make the code
support the maximum potential deployments while the opposite objective
is in the packaging which is very specific to platform and so should
just pick one, as the packagers are trying to minimise support costs and
diversity.

I expect finally we have a make target such as:

	make rpm

Autotools (or cmake which I know less) is very very rich and flexible
having been designed to address the differences between all the unix
systems and they are rather naturally understood as programming
languages by developers trying to support many target systems.

To me a distribution has the job of taking all products in them and
making them as consistent as possible to give their users less
difficulty managing them. I see packages having conventions and macros
in the case of rpm and functions in the case of deb that are very
distribution specific and help make all packages behave the same in a
given distribution.

For a distribution: varnish (for example) and ceph should be similar to
start, stop and upgrade.

For example I expect distributions to upgrade, and transition from one
version to another, in a way defined by the distribution. With SUSE
debian and scientific linux (RHEL as I am used to it) this means that
services should never be running a different version from the version
installed.

I have no desire to make varnish (as an example) and ceph upgrade the
same way at a code level but the bits end users see should be similar so
admins dont need to write many configuration management scripts for
their data centre for every service.


Summary
-------

Its your decision (1) or (2)

I think (2)

I think you should demand that this patch cannot go in until we have a
top level configure switches  such as:

  --with-systemd
  --with-sysv
  --with-upstart

Nearly all paths used by the spec.in file must be templates from the
configure script.

We can then have all the logic for different distributions set once and
only once in the initial configure stage, and you can delegate all the
boring details of differences between SUSE Redhat and Debian to the
packaging files.

You could even demand that products such as upstart and systemd that can
emulate sysV can do transition between one and other transparently
relatively easily, but this is final acceptance criteria.

Best regards

Owen

osynge commented:
discussed this at ceph standup and concluded we will discuss policy on the mailing list.

osynge commented:
I am going to close this until I have fixed the bugs in it.

osynge commented:
I think its fit to reach DNM stage

osynge commented:
Oh I cant reopen.

ktdreyer commented:
GitHub's UI is telling me "there is already an open pull request SUSE:wip-systemd_autotools_install_unit"

osynge commented:
replaced by https://github.com/ceph/ceph/pull/5029
